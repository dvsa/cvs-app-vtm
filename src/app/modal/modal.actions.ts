import { Action } from '@ngrx/store';
import { ModalState } from './modal.reducer';

export enum EModalStateActions {
  LoadModal = '[Load] Modal',
  ResetModal = '[Reset] Modal',
  RedirectModal = '[Redirect] Modal'
}

export class LoadModal implements Action {
  public readonly type = EModalStateActions.LoadModal;
  constructor(public payload: ModalState) {}
}

export class ResetModal implements Action {
  public readonly type = EModalStateActions.ResetModal;
  constructor(public urlToRedirect?: string) {}
}

export class RedirectModal implements Action {
  public readonly type = EModalStateActions.RedirectModal;
  constructor(public urlToRedirect?: string) {}
}

export type ModalStateActions = LoadModal | ResetModal;
