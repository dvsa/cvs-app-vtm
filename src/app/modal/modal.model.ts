import { APP_MODALS } from '../app.enums';

export interface ModalContent {
  modal: APP_MODALS;
  status: boolean;
  payload?: any;
  urlToRedirect: string;
}
