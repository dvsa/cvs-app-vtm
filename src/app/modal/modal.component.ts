import { Component, OnInit, ChangeDetectionStrategy, Inject } from '@angular/core';
import { APP_MODALS } from '../app.enums';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
import { ModalContent } from './modal.model';
import { Store } from '@ngrx/store';
import { IAppState } from '@app/store/state/app.state';
import { ResetModal } from './modal.actions';
import { ModalContainerComponent } from './modal.container.component';

@Component({
  selector: 'vtm-modal',
  template: `
    <ng-container *ngIf="currentModal.modal === 'lose-changes'">
      <vtm-lose-changes (okCancelAction)="okCancelAction($event)"></vtm-lose-changes>
    </ng-container>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ModalComponent implements OnInit {
  public currentModal: ModalContent;

  constructor(
    @Inject(MAT_DIALOG_DATA) data,
    public dialogRef: MatDialogRef<ModalComponent>,
    private store: Store<IAppState>
  ) {
    this.currentModal = {
      modal: data.currentModal,
      urlToRedirect: data.currentRoute
    } as ModalContent;
  }

  ngOnInit() {}

  okCancelAction(isOk: boolean) {
    this.handleModal({ ...this.currentModal, status: isOk });
    this.dialogRef.close();
  }

  handleModal(modalContent: ModalContent) {
    switch (modalContent.modal) {
      case APP_MODALS.LOSE_CHANGES:
        const redirectUrl = modalContent.status ? modalContent.urlToRedirect : '';
        this.store.dispatch(new ResetModal(redirectUrl));
        return;
    }
  }
}
