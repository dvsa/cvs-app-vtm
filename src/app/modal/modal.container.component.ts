import { Component, OnInit, ChangeDetectionStrategy, Input } from '@angular/core';
import { MatDialogConfig, MatDialog } from '@angular/material/dialog';

import { APP_MODALS } from '../app.enums';
import { IAppState } from '@app/store/state/app.state';
import { Store } from '@ngrx/store';
import { getCurrentModalState } from './modal.selectors';
import { Observable } from 'rxjs';
import { ModalComponent } from './modal.component';
import { ModalState } from './modal.reducer';

@Component({
  selector: 'vtm-modal-container',
  template: `
    <ng-content></ng-content>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ModalContainerComponent implements OnInit {
  @Input() currentModal$: Observable<ModalState>;
  constructor(private dialog: MatDialog, private store: Store<IAppState>) {}

  ngOnInit() {
    this.currentModal$ = this.store.select(getCurrentModalState);

    this.currentModal$.subscribe((modal: ModalState) => {
      if (modal.currentModal !== APP_MODALS.NONE) {
        const dialogConfig = new MatDialogConfig();

        dialogConfig.disableClose = true;
        dialogConfig.autoFocus = true;
        dialogConfig.width = '45vw';
        dialogConfig.data = {
          currentModal: modal.currentModal,
          currentRoute: modal.currentRoute
        };
        this.dialog.open(ModalComponent, dialogConfig);
      } else {
        this.dialog.closeAll();
      }
    });
  }
}
