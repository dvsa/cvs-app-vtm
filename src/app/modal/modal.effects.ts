import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { Observable, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { Action } from '@ngrx/store';
import { Effect, Actions, ofType } from '@ngrx/effects';

import { VIEW_STATE } from '@app/app.enums';
import { ResetModal, RedirectModal, EModalStateActions } from './modal.actions';
import {
  SetViewState
} from '@app/store/actions/VehicleTechRecordModel.actions';
import { SetAppFormPristine } from '@app/store/actions/app-form-state.actions';

@Injectable()
export class ModalEffects {
  @Effect()
  resetModal$: Observable<Action> = this.actions$.pipe(
    ofType<ResetModal>(EModalStateActions.ResetModal),
    map((action) => action.urlToRedirect),
    switchMap((urlToRedirect?: string) => {
      const actions = [];
      if (urlToRedirect) {
        actions.push(new RedirectModal(urlToRedirect));
        actions.push(new SetAppFormPristine());
        actions.push(new SetViewState(VIEW_STATE.VIEW_ONLY));
      }
      return actions;
    })
  );

  @Effect({ dispatch: false })
  redirectModal$: Observable<Action> = this.actions$.pipe(
    ofType<RedirectModal>(EModalStateActions.RedirectModal),
    map((action) => action.urlToRedirect),
    switchMap((urlToRedirect?: string) => {
      if (urlToRedirect) {
          this.router.navigateByUrl(urlToRedirect);
      }
      return of(undefined);
    })
  );

  constructor(private actions$: Actions, private router: Router) {}
}
