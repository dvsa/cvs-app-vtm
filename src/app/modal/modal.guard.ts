import { ActivatedRouteSnapshot, CanActivate, CanDeactivate } from '@angular/router';
import { Observable, of } from 'rxjs';
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { IAppState } from '@app/store/state/app.state';
import { map, switchMap } from 'rxjs/operators';
import { getAppFormState } from '@app/store/selectors/app-form-state.selectors';
import { LoadModal } from './modal.actions';
import { APP_MODALS } from '@app/app.enums';
import { TechnicalRecordsContainer } from '@app/technical-record/technical-record.container';
import { TechnicalRecordComponent } from '@app/technical-record/technical-record.component';

@Injectable({ providedIn: 'root' })
export class ModalGuard implements CanDeactivate<TechnicalRecordsContainer> {
  constructor(private store: Store<IAppState>) {}

  isAppFormDirty(): Observable<boolean> {
    return this.store.pipe(
      select(getAppFormState),
      map((pristine) => {
        return pristine;
      })
    );
  }

  canDeactivate(
    component: TechnicalRecordsContainer,
    currentRoute: ActivatedRouteSnapshot,
    currentState: import('@angular/router').RouterStateSnapshot,
    nextState?: import('@angular/router').RouterStateSnapshot
  ): boolean | Observable<boolean> | Promise<boolean> {
    return this.isAppFormDirty().pipe(
      switchMap((pristine) => {
        if (!pristine) {
          this.store.dispatch(
            new LoadModal({
              currentModal: APP_MODALS.LOSE_CHANGES,
              currentRoute: getResolvedUrl(currentRoute)
            })
          );
        }
        return of(pristine);
      })
    );
  }

  // canActivate(route: ActivatedRouteSnapshot): Observable<boolean> {
  //   return this.isAppFormDirty().pipe(
  //     switchMap((pristine) => {
  //       if (!pristine) {
  //         this.store.dispatch(
  //           new LoadModal({
  //             currentModal: APP_MODALS.LOSE_CHANGES,
  //             currentRoute: getResolvedUrl(route)
  //           })
  //         );
  //       }
  //       return of(pristine);
  //     })
  //   );
  // }
}

function getResolvedUrl(route: ActivatedRouteSnapshot): string {
  return route.pathFromRoot
    .map((v) => v.url.map((segment) => segment.toString()).join('/'))
    .join('/');
}
