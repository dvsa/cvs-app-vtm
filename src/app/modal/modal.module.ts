import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ModalContainerComponent } from './modal.container.component';
import { LoseChangesComponent } from './components/lose-changes/lose-changes.component';
import { ModalComponent } from './modal.component';
import { ModalEffects } from './modal.effects';
import { EffectsModule } from '@ngrx/effects';

@NgModule({
  imports: [
    CommonModule,
    EffectsModule.forFeature([ModalEffects]),
  ],
  declarations: [ModalContainerComponent, ModalComponent, LoseChangesComponent],
  entryComponents: [ModalComponent, LoseChangesComponent],
  exports: [ModalContainerComponent],
})
export class ModalModule { }
