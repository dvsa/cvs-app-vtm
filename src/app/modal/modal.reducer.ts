import { APP_MODALS } from '@app/app.enums';
import { ModalStateActions, EModalStateActions } from './modal.actions';

export interface ModalState {
  currentModal: APP_MODALS;
  currentRoute: string;
}

export const initialModalState: ModalState = {
  currentModal: APP_MODALS.NONE,
  currentRoute: ''
};

export function ModalReducer(
  state: ModalState = initialModalState,
  action: ModalStateActions
): ModalState {
  switch (action.type) {
    case EModalStateActions.LoadModal: {
      return {
        ...state,
        currentModal: action.payload.currentModal,
        currentRoute: action.payload.currentRoute
      };
    }

    case EModalStateActions.ResetModal: {
      return {
        ...state,
        currentModal: APP_MODALS.NONE,
        currentRoute: action.urlToRedirect
      };
    }

    default:
      return state;
  }
}
