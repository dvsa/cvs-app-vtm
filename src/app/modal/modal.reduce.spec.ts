import { Action } from '@ngrx/store';

import { ModalState, initialModalState, ModalReducer } from './modal.reducer';
import { LoadModal, ResetModal } from './modal.actions';
import { APP_MODALS } from '../app.enums';

describe('ModalReducer', () => {
  let state: ModalState;
  let resultState: ModalState;
  let action: any;

  describe('when initialised', () => {
    beforeEach(() => {
      action = {} as Action;
      const initState: ModalState = initialModalState;
      resultState = ModalReducer(initState, action);
    });

    it('should set the default state', () => {
      expect(resultState).toMatchSnapshot();
    });
  });

  describe('LoadModal', () => {
    it('should set the error(s) to the state', () => {
      action = new LoadModal({ currentModal: APP_MODALS.NONE, currentRoute: 'someRoute' });
      state = {
        currentModal: APP_MODALS.NONE,
        currentRoute: ''
      };

      resultState = ModalReducer(state, action);

      expect(resultState).toMatchSnapshot();
    });
  });

  describe('ResetModal', () => {
    it('should clear error(s) from the state', () => {
      action = new ResetModal('someUrl');
      state = {
        currentModal: APP_MODALS.LOSE_CHANGES,
        currentRoute: ''
      };

      resultState = ModalReducer(state, action);

      expect(resultState).toMatchSnapshot();
    });
  });
});
