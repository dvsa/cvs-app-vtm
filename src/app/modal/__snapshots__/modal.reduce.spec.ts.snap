// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalReducer LoadModal should set the error(s) to the state 1`] = `
Object {
  "currentModal": "none",
  "currentRoute": "someRoute",
}
`;

exports[`ModalReducer ResetModal should clear error(s) from the state 1`] = `
Object {
  "currentModal": "none",
  "currentRoute": "someUrl",
}
`;

exports[`ModalReducer when initialised should set the default state 1`] = `
Object {
  "currentModal": "none",
  "currentRoute": "",
}
`;
